### PR TITLE
Add support for empty lists

### PR DESCRIPTION
**Pull request by jazg**

### CODE REVIEW

File: list_test.go [hunk 1]
```diff
@@ -6,14 +6,13 @@ import (
...
 	"github.com/renproject/pack/packutil"
-	"github.com/renproject/surge"
 	"github.com/renproject/surge/surgeutil"
 
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
 )
 
-var _ = Describe("List", func() {
+var _ = FDescribe("List", func() {
```
Comment (loongy) [inline list_test.go:15]: @jazg you will need to remove this focus

Comment (loongy) [inline list_test.go:15]:
```suggestion
var _ = Describe("List", func() {
```

File: list_test.go [hunk 2]
```diff
@@ -66,21 +65,4 @@ var _ = Describe("List", func() {
...
-
-	Context("when unmarshaling an empty list", func() {
-		It("should return an error", func() {
-			list := pack.List{}
-			bytes, err := surge.ToBinary(list)
-			Expect(err).ToNot(HaveOccurred())
-			err = surge.FromBinary(&list, bytes)
-			Expect(err).To(HaveOccurred())
-		})
-	})
```
Comment (loongy) [inline list_test.go:85]: We should keep these test cases, but explicitly check that they do *not* result in errors, but do result in Lists with the Nil element type.

File: nil.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	"reflect"
+)
+
+// Nil represents a nil value.
+type Nil struct{}
+
+// NewNil wraps an existing raw struct.
+func NewNil(x struct{}) Nil {
+	return Nil(x)
+}
```
Comment (loongy) [inline nil.go:14]: Why does it wrap an existing empty struct? Why not:
```suggestion
// NewNil wraps an existing raw struct.
func NewNil() Nil {
	return Nil{}
}
```

File: nil.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+// Type returns nil type.
+func (x Nil) Type() Type {
+	return typeNil{}
+}
+
+// Equal returns true when x is equal to y. Otherwise, it returns false.
+func (x Nil) Equal(y Nil) bool {
+	return x == y
```
Comment (loongy) [inline nil.go:23]:
```suggestion
	return true
```

Comment (loongy) [reply nil.go:23]: Nil values are always equal to each other, because there is no "room" for variability.

File: nil.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+// String returns "" for nil values.
+func (x Nil) String() string {
+	return ""
+}
+
+// Generate a random nil value. This method is implemented for use in quick
+// tests. See https://golang.org/pkg/testing/quick/#Generator for more
+// information.
+func (Nil) Generate(r *rand.Rand, size int) reflect.Value {
+	return reflect.ValueOf(NewNil(struct{}{}))
```
Comment (loongy) [inline nil.go:63]:
```suggestion
	return reflect.ValueOf(NewNil())
```

File: nil_test.go [hunk 6]
```diff
@@ -0,0 +1 @@
+package pack
```
Comment (loongy) [inline nil_test.go:1]:
```suggestion
package pack_test
```
Some basic tests should be added.

File: type.go [hunk 7]
```diff
@@ -29,6 +29,73 @@ type Type interface {
...
+	return KindNil
+}
+
+func (t typeNil) Equals(other Type) bool {
+	_, ok := other.(typeNil)
+	return ok
+}
+
+func (typeNil) UnmarshalValue(buf []byte, rem int) (Value, []byte, int, error) {
+	value := Nil(struct{}{})
```
Comment (loongy) [inline type.go:44]:
```suggestion
	value := Nil{}
```

File: type.go [hunk 8]
```diff
@@ -29,6 +29,73 @@ type Type interface {
...
+	buf, rem, err := value.Unmarshal(buf, rem)
+	return value, buf, rem, err
+}
+
+func (typeNil) UnmarshalValueJSON(data []byte) (Value, error) {
+	value := Nil(struct{}{})
```
Comment (loongy) [inline type.go:50]:
```suggestion
	value := Nil{}
```

File: nil.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+// MarshalJSON marshals the nil value to JSON.
+func (x Nil) MarshalJSON() ([]byte, error) {
+	return nil, nil
+}
+
+// UnmarshalJSON unmarshals the nil value from JSON.
+func (x *Nil) UnmarshalJSON(data []byte) error {
+	x = (*Nil)(&struct{}{})
+	return nil
+}
```
Comment (loongy) [inline nil.go:52]: Might be worth explicitly marshalling to/from JSON using the `nil` JSON value. That way we are able to distinguish between "not there" and "there, but nil".

**Review by loongy — CHANGES_REQUESTED** [review]

File: nil_test.go (hunk 6 above)
Comment (jazg) [reply nil_test.go:1]: This was just a draft PR :)

File: list.go [hunk 10]
```diff
@@ -61,7 +63,7 @@ func (v *List) Unmarshal(buf []byte, rem int) ([]byte, int, error) {
 		return buf, rem, err
 	}
 	if len(v.Elems) == 0 {
-		return buf, rem, fmt.Errorf("cannot construct list with no elements")
+		v.T = typeNil{}
```
Comment (loongy) [inline list.go:66]: This actually makes me realise there is a bug for the case when `len(v.Elems) > 0`. At the moment, the `v.T` would be left uninitialised, but it should be set to the type of `v.Elems[0]` and all elements should be checked to make sure they are consistent with that type.

**Review by loongy — CHANGES_REQUESTED** [review]
Everything is looking good, but I found one bug (not with your changes, but made more obvious by your changes).

**Comment by jazg** [comment]
This PR is mostly complete but there are some open issues:
1. Struct binary marshal/unmarshal tests (previously didn't exist) are failing.
2. List JSON marshal/unmarshal tests (previously didn't exist) are failing.
3. We could introduce an `EncodeType` function and make use of it when handling empty lists in the `Encode` function, for a performance optimisation.

**Review by loongy — APPROVED** [review]

**Pull request merged**